File: src/main/java/chat/teco/tecochat/chat/domain/chat/QuestionAndAnswers.java
package chat.teco.tecochat.chat.domain.chat;

import static jakarta.persistence.CascadeType.ALL;
import static jakarta.persistence.FetchType.LAZY;

import jakarta.persistence.Embeddable;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OneToMany;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Embeddable
public class QuestionAndAnswers {

    @OneToMany(fetch = LAZY, cascade = ALL, orphanRemoval = true)
    @JoinColumn(name = "chat_id")
    private List<QuestionAndAnswer> questionAndAnswers = new ArrayList<>();

    protected QuestionAndAnswers() {
    }

    public QuestionAndAnswers(QuestionAndAnswer... questionAndAnswers) {
        this(Arrays.asList(questionAndAnswers));
    }

    public QuestionAndAnswers(List<QuestionAndAnswer> questionAndAnswers) {
        this.questionAndAnswers.addAll(questionAndAnswers);
    }

    public void add(QuestionAndAnswer questionAndAnswer) {
        questionAndAnswers.add(questionAndAnswer);
    }

    public QuestionAndAnswers last3QuestionAndAnswers() {
        int size = questionAndAnswers.size();
        if (size < 3) {
            return this;
        }
        return new QuestionAndAnswers(questionAndAnswers.subList(size - 3, size));
    }

    public List<Message> messagesWithSettingMessage(SettingMessage settingMessage) {
        List<Message> result = new ArrayList<>();
        result.add(settingMessage);
        for (QuestionAndAnswer qna : questionAndAnswers) {
            result.add(qna.question());
            result.add(qna.answer());
        }
        return result;
    }

    public List<QuestionAndAnswer> questionAndAnswers() {
        return new ArrayList<>(questionAndAnswers);
    }
}

File: src/main/kotlin/chat/teco/tecochat/domain/chat/QuestionAndAnswers.kt
package chat.teco.tecochat.domain.chat

import jakarta.persistence.CascadeType
import jakarta.persistence.Embeddable
import jakarta.persistence.FetchType
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToMany

@Embeddable
class QuestionAndAnswers(
    @OneToMany(fetch = FetchType.LAZY, cascade = [CascadeType.ALL], orphanRemoval = true)
    @JoinColumn(name = "chat_id")
    val questionAndAnswers: MutableList<QuestionAndAnswer> = ArrayList(),
) {

    fun add(questionAndAnswer: QuestionAndAnswer) {
        questionAndAnswers.add(questionAndAnswer)
    }

    fun last3QuestionAndAnswers(): QuestionAndAnswers {
        val size = questionAndAnswers.size
        return if (size < 3) {
            this
        } else QuestionAndAnswers(questionAndAnswers.subList(size - 3, size))
    }

    fun messagesWithSettingMessage(settingMessage: SettingMessage): List<Message> {
        val result: MutableList<Message> = ArrayList()
        result.add(settingMessage)
        for (qna in questionAndAnswers) {
            result.add(qna.question)
            result.add(qna.answer)
        }
        return result
    }

    fun questionAndAnswers(): List<QuestionAndAnswer> {
        return ArrayList(questionAndAnswers)
    }
}
